# origin: Debian (#272719)
# cset: 1.1722.16.9 (2.5/2.6)
# inclusion: ipsec is not in upstream
# description: resolve bug that prevents spdadd correctly effecting spddelete
# revision date: 2004-09-24

#Reported to debian-kernel by Teddy Hogeborn <teddy@fukt.bth.se>
#Bug#272719
#
#ChangeSet1.1722.16.9
# This is a BitKeeper generated diff -Nru style patch.
#
# ChangeSet
#   2004/05/25 13:09:27-07:00 sxking@qwest.net 
#   [IPSEC]: Fix buglet in AF_KEY spddelete
#   
#   When trying to spddelete individual entries using setkey, spddelete always 
#   fails.  The culprit is in net/af_key.c; spdadd sets the family field of the 
#   selector when creating an entry, but spddelete doesn't when building a 
#   selector to match for xfrm_policy_bysel.  Trivial fix is to have spddelete 
#   set the family field in the selector in same way spdadd does.
# 
# net/key/af_key.c
#   2004/05/25 13:09:13-07:00 sxking@qwest.net +1 -1
#   [IPSEC]: Fix buglet in AF_KEY spddelete
#   
#   When trying to spddelete individual entries using setkey, spddelete always 
#   fails.  The culprit is in net/af_key.c; spdadd sets the family field of the 
#   selector when creating an entry, but spddelete doesn't when building a 
#   selector to match for xfrm_policy_bysel.  Trivial fix is to have spddelete 
#   set the family field in the selector in same way spdadd does.
# 
diff -Nru a/net/key/af_key.c b/net/key/af_key.c
--- a/net/key/af_key.c	2004-09-21 19:32:47 -07:00
+++ b/net/key/af_key.c	2004-09-21 19:32:47 -07:00
@@ -1976,7 +1976,7 @@
 	memset(&sel, 0, sizeof(sel));
 
 	sa = ext_hdrs[SADB_EXT_ADDRESS_SRC-1], 
-	pfkey_sadb_addr2xfrm_addr(sa, &sel.saddr);
+	sel.family = pfkey_sadb_addr2xfrm_addr(sa, &sel.saddr);
 	sel.prefixlen_s = sa->sadb_address_prefixlen;
 	sel.proto = pfkey_proto_to_xfrm(sa->sadb_address_proto);
 	sel.sport = ((struct sockaddr_in *)(sa+1))->sin_port;
